pkg/bootstrap/bootstrap.go

```diff
@@ -220,93 +278,210 @@ func extractToDir(dir, prefix string, img v1.Image, imageName string) error {
+		if _, err := os.Stat(tempSource); err != nil {
+			errs = append(errs, err)
+			continue
+		}
 
-	var errs []error
-	for _, file := range files {
-		src := filepath.Join(tempDir, file.Name())
-		dst := filepath.Join(dir, file.Name())
-		if err := os.Rename(src, dst); os.IsExist(err) {
-			// Can't rename because dst already exists, remove it...
-			if err = os.RemoveAll(dst); err != nil {
-				errs = append(errs, errors2.Wrapf(err, "failed to remove %q", dst))
-				continue
+		logrus.Debugf("Moving extracted content from %s to %s", tempSource, dest)
+
+		// Ensure the parent directory exists before renaming into the target location
+		destParent := filepath.Dir(dest)
+		if _, err := os.Stat(destParent); err != nil {
+			if err := os.MkdirAll(destParent, 0755); err != nil {
+				return err
 			}
-			// ...then try renaming again
-			if err = os.Rename(src, dst); err != nil {
-				errs = append(errs, errors2.Wrapf(err, "failed to rename %q to %q", src, dst))
+		}
+
+		// Try to rename the temp dir into its target location.
+		if err := os.Rename(tempSource, dest); err == nil {
+			// Successfully renamed into place, nothing else to do.
+			continue
+		} else if !os.IsExist(err) {
+			// Failed to rename, but not because the destination already exists.
+			errs = append(errs, err)
+			continue
+		}
+
+		// Target directory already exists (got ErrExist above), fall back list/rename files into place.
+		files, err := ioutil.ReadDir(tempSource)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		for _, file := range files {
+			src := filepath.Join(tempSource, file.Name())
+			dst := filepath.Join(dest, file.Name())
+			if err := os.Rename(src, dst); os.IsExist(err) {
+				// Can't rename because dst already exists, remove it...
+				if err = os.RemoveAll(dst); err != nil {
+					errs = append(errs, errors.Wrapf(err, "failed to remove %s", dst))
+					continue
+				}
+				// ...then try renaming again
+				if err = os.Rename(src, dst); err != nil {
+					errs = append(errs, errors.Wrapf(err, "failed to rename %s to %s", src, dst))
+				}
+			} else if err != nil {
+				// Other error while renaming src to dst.
+				errs = append(errs, errors.Wrapf(err, "failed to rename %s to %s", src, dst))
 			}
-		} else if err != nil {
-			// Other error while renaming src to dst.
-			errs = append(errs, errors2.Wrapf(err, "failed to rename %q to %q", src, dst))
 		}
 	}
+
 	return merr.NewErrors(errs...)
 }
 
-// preloadBootstrapImage attempts return an image named imageName from a tarball
+// preloadBootstrapFromRuntime tries to load the runtime image from tarballs, using both the
+// default registry, and the user-configured registry (on the off chance they've retagged the
+// images in the tarball to match their private registry).
+func preloadBootstrapFromRuntime(dataDir string, resolver *images.Resolver) (v1.Image, error) {
+	var refs []name.Reference
+	runtimeRef := resolver.MustGetReference(images.Runtime)
```

Comment:
The return signature of `preloadBootstrapFromRuntime` includes an error so we should return it instead of delegating to `logrus.Fatalf` in the `MustGetReference` wrapper.